test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -362,44 +361,112 @@ class test_ucp_sockaddr : public ucp_test {
-            ASSERT_UCS_OK(send_status);
+            EXPECT_UCS_OK(send_status);
         }
 
+        return true;
+    }
+
+    void* send(entity& from, const void *contig_buffer, size_t length,
+               send_recv_type_t send_type, ucp_send_nbx_callback_t cb,
+               void *user_data  )
+    {
+        ucp_request_param_t params = {};
+
+        params.op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK |
+                              UCP_OP_ATTR_FIELD_USER_DATA;
+        params.cb.send      = cb;
+        params.user_data    = user_data;
+
+        void *send_req = NULL;
+        if (send_type == SEND_RECV_TAG) {
+            send_req = ucp_tag_send_nbx(from.ep(), contig_buffer, length, 1,
+                                        &params);
+        } else if (send_type == SEND_RECV_STREAM) {
+            send_req = ucp_stream_send_nbx(from.ep(), contig_buffer, length,
+                                           &params);
+        } else {
+            UCS_TEST_ABORT("unsupported communication type " << send_type);
+        }
+
+        return send_req;
+    }
+
+    void* recv(entity& to, void *contig_buffer, size_t length,
+               ucp_tag_recv_nbx_callback_t cb, void *user_data)
+    {
+        ucp_request_param_t params = {};
+
+        params.op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK |
+                              UCP_OP_ATTR_FIELD_USER_DATA;
+        params.user_data    = user_data;
+        params.cb.recv      = cb;
+        return ucp_tag_recv_nbx(to.worker(), contig_buffer, length, 1, 0,
+                                &params);
+    }
+
+    void* recv(entity& to, void *contig_buffer, size_t length,
+               ucp_stream_recv_nbx_callback_t cb, void *user_data)
+    {
+        ucp_request_param_t params = {};
+
+        params.op_attr_mask   = UCP_OP_ATTR_FIELD_CALLBACK |
+                                UCP_OP_ATTR_FIELD_USER_DATA |
+                                UCP_OP_ATTR_FIELD_FLAGS;
+        params.flags          = UCP_STREAM_RECV_FLAG_WAITALL;
+        params.user_data      = user_data;
+        params.cb.recv_stream = cb;
+
+        ucs_time_t deadline = ucs::get_deadline();
+        ucp_stream_poll_ep_t poll_eps;
+        ssize_t ep_count;
+        do {
+            progress();
+            ep_count = ucp_stream_worker_poll(to.worker(), &poll_eps, 1, 0);
+        } while ((ep_count == 0) && (ucs_get_time() < deadline));
+        EXPECT_EQ(1, ep_count);
+        EXPECT_EQ(to.ep(), poll_eps.ep);
+        EXPECT_EQ(&to, poll_eps.user_data);
+
+        size_t recv_length;
+        return ucp_stream_recv_nbx(to.ep(), contig_buffer, length,
+                                   &recv_length, &params);
+    }
+
+    void send_recv(entity& from, entity& to, send_recv_type_t send_recv_type,
+                   bool wakeup, ucp_test_base::entity::listen_cb_type_t cb_type)
+    {
+        const uint64_t send_data = ucs_generate_uuid(0);
+        void *send_req           = send(from, &send_data, sizeof(send_data),
```

Comment:
better not init inline since it's not a "trivial" initialization

Reply:
fixed